### PR TITLE
[clang] Respect the lifetimebound in assignment operator.

### LINKED ISSUE #106372: Respect the lifetimebound in assignment operator.
https://godbolt.org/z/43nf1v349

```
#include <vector>
#include <string>

template <typename T>
class  StatusOr {
public:
  template <typename U = T>
  StatusOr(U&& t [[clang::lifetimebound]]);

  template <typename U = T>
  StatusOr& operator=(U&& v [[clang::lifetimebound]]);
};

void test() {
  StatusOr<std::string_view> foo = std::string(); // #1 good, clang emits dangling diagnostic
  foo = std::string(); // #2, no dangling diagnostic
}
```

In the above example, clang doesn't catch the case 2, it would be nice to catch this case. We can assume that all assignment operators with a typical return type (`const reference&`) always return *this.

### PR DESCRIPTION
**Pull request by hokein**
Fixes #106372

### CODE REVIEW

File: clang/lib/Sema/CheckExprLifetime.cpp [hunk 1]
```diff
@@ -968,6 +972,23 @@ static bool pathOnlyHandlesGslPointer(IndirectLocalPath &Path) {
   return false;
 }
 
+static bool isAssginmentOperatorLifetimeBound(CXXMethodDecl *CMD) {
+  if (!CMD)
+    return false;
+  assert(CMD->getOverloadedOperator() == OverloadedOperatorKind::OO_Equal);
```
Comment (Xazax-hun) [inline clang/lib/Sema/CheckExprLifetime.cpp:978]: What about compound assignment operators? Do you plan to support those in a follow-up?

That being said, I'd love to see a test case just to make sure this assert is not triggered.

File: clang/test/SemaCXX/attr-lifetimebound.cpp [hunk 2]
```diff
@@ -287,3 +287,18 @@ std::span<int> test2() {
...
+class [[gsl::Pointer]] FooView {};
+
+template <typename T>
+struct StatusOr {
+  template <typename U = T>
+  StatusOr& operator=(U&& v [[clang::lifetimebound]]);
+};
+
+void test(StatusOr<FooView> foo) {
+  foo = Foo(); // expected-warning {{object backing the pointer foo will be destroyed at the end}}
```
[the remark this answers is not in the document]
Comment (Xazax-hun) [reply clang/test/SemaCXX/attr-lifetimebound.cpp:302]: Oh, never mind. Actually, in that specialization people would not mark the argument of `operator=` lifetimebound. This makes total sense. Disregard this.

**Review by Xazax-hun — APPROVED** [review]
After a couple more tests LGTM!

**Review by hokein — COMMENTED** [review]
Thanks for the quick review.

File: clang/lib/Sema/CheckExprLifetime.cpp [hunk 3]
```diff
@@ -326,24 +326,11 @@ static bool shouldTrackFirstArgument(const FunctionDecl *FD) {
...
-  }
-
-  // Assume that all assignment operators with a "normal" return type return
-  // *this, that is, an lvalue reference that is the same type as the implicit
-  // object parameter (or the LHS for a non-member operator$=).
+// Return true if this is an "normal" assignment operator.
+// We assuments that a normal assingment operator always returns *this, that is,
+// an lvalue reference that is the same type as the implicit object parameter
+// (or the LHS for a non-member operator$=).
+static bool isNormalAsisgnmentOperator(const FunctionDecl *FD) {
```
Comment (aeubanks) [inline clang/lib/Sema/CheckExprLifetime.cpp:333]: nit: "Assignment" is spelled incorrectly here and also below

Comment (hokein) [reply clang/lib/Sema/CheckExprLifetime.cpp:333]: Thanks, have fixed in https://github.com/llvm/llvm-project/commit/87b4b648585f69a2ea148e86543aa31474e59acd.

**Pull request merged**